keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,273 @@
+This doc proposes a new EndpointSlice API aiming to replace Core/V1 Endpoints API for most internal consumers, including kube-proxy.
+The new EndpointSlice API aims to address existing problems as well as leaving room for future extension.
+
+
+## Motivation
+
+In the current Endpoints API, one object instance contains all the individual endpoints of a service. Whenever a single pod in a service is added/updated/deleted, the whole Endpoints object (even when the other endpoints didn't change) is re-computed, written to storage (etcd) and sent to all watchers (e.g. kube-proxy). This leads to 2 major problems:
+
+- Storing multiple megabytes of endpoints puts strain on multiple parts of the system due to not having a paging system and a monolithic watch/storage design. [The max number of endpoints is bounded by the K8s storage layer (etcd)][max-object-size]], which has a hard limit on the size of a single object (1.5MB by default). That means attempts to write an object larger than the limit will be rejected. Additionally, there is a similar limitation in the watch path in Kubernetes apiserver. For a K8s service, if its Endpoints object is too large, endpoint updates will not be propagated to kube-proxy(s), and thus iptables/ipvs won’t be reprogrammed.
+- [Performance degradation in large k8s deployments.][perf-degrade] Not being able to efficiently read/update individual endpoint changes can lead to (e.g during rolling upgrade of a service) endpoints operations that are quadratic in the number of its elements. If one consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the quadratic traffic gets multiplied further with number of watches (usually equal to #nodes in the cluster).
+
+The new EndpointSlice API aims to address existing problems as well as leaving room for future extension.
+
+
+### Goal
+
+- Support tens of thousands of backend endpoints in a single service on cluster with thousands of nodes.
+- Leave room for foreseeable extension:
+  - Support multiple IPs per pod
+  - More endpoint states than Ready/NotReady
+  - Dynamic endpoint subsetting
+  
+### Non-Goal
+- Change functionality provided by K8s V1 Service API.
+- Provide better load balancing for K8s service backends.
+
+## Proposal
+
+### EndpointSlice API
+The following new EndpointSlice API will be added to the networking API group. 
+
+```
+type EndpointSlice struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+        Spec EndpointSliceSpec
+}
+
+type EndpointSliceSpec struct {
+        Endpoints []Endpoint
+        Ports []EndpointPort
+}
+
+type EndpointPort struct {
+        // The name of this port (corresponds to ServicePort.Name).
+        // Must be a DNS_LABEL.
+        // Optional only if one port is defined.
```

Comment:
This conditional "requiredness" is a little confusing. Can we group the fields in a different way?

Reply:
Sadly, this is how the legacy API is written.  But we should do better.

Reply:
Yeah, I agree it is confusing. But it is currently everywhere, from pod spec, service spec all the way to endpoints.

Reply:
Since this is a new resource we can change the rules.   We should think of this from the POV of the api machinery and the use-cases.  Do we want to allow people to patch these lists or do we see them as atomic?  I think addresses certainly wants to be patched, but ports I am less 100% sure.  

To be patchable we need a good merge key, so optional values are not great for that.

We could argue that the merge key is "port number + protocol", and name is just a decoration.  Even then, sometimes optional is blech.  Also, protocol is optional-with-default, which makes the machinery ugly.

We could argue that the merge key is name, and that "" is a valid name.  We'd have to look at places that do named-port lookups against endpoints and make sure that "" is allowed.

That still leaves a bit of ugly in that we really *should* be able to use the same name across protocols (e.g. DNS has them as different fields in SRV) but that, again, is made ugly by protocol-defaulting.

So I think my position is that `name` is the (required) merge key and "" is a valid name (not omitempty).  Looking at @apelisse for more formal guidance.  It seems to me that sometimes-optional should be described somewhere as an anti-pattern.

Reply:
> To be patchable we need a good merge key, so optional values are not great for that.

Yes, merge key(s) should not be optional.